route_guide/route_guide_client.cc

```diff
@@ -23,238 +23,347 @@
+                  const Feature& f =
+                      feature_list_[feature_distribution(generator)];
+                  std::cout << "Visiting point "
+                            << f.location().latitude() / kCoordFactor_
+                            << ", "
+                            << f.location().longitude() / kCoordFactor_
+                            << std::endl;
+                  call.Write(f.location());
+                  std::this_thread::sleep_for(std::chrono::milliseconds(
+                      delay_distribution(generator)));
+                }
+                // When writes are done we should call WritesDone in order
+                // to finish our call.
+                call.WritesDone();
+              })
+
+              // The .body() callback is invoked every time the server
+              // sends a response.
+              .body([](auto&& data, auto& call, auto&& response) {
+                // We should check if the response is valid from the server.
+                if (!response) {
+                  return;
+                }
+                data.stats.Swap(response.get());
+                std::cout << "Finished trip with "
+                          << data.stats.point_count() << " points\n"
+                          << "Passed " << data.stats.feature_count()
+                          << " features\n"
+                          << "Travelled " << data.stats.distance()
+                          << " meters\n"
+                          << "It took " << data.stats.elapsed_time()
+                          << " seconds" << std::endl;
+              })
+              // .finished callback is invoked when the call is finished.
+              // In this callback we just start our continuation.
+              .finished([](
+                            auto&& data,
+                            auto& continuation,
+                            auto&& response) {
+                continuation.Start(grpc::Status::OK);
+              });
   }
 
-  void RouteChat() {
-    std::thread writer;
-    Notification<Status> done;
-    client_->Call<RouteGuide, Stream<RouteNote>, Stream<RouteNote>>(
-        "RouteChat",
-        [&](auto&& call, bool ok) {
-          call->OnFinished([&](auto*, const Status& status) {
-            done.Notify(status);
-          });
-          if (!ok) {
-            call->context()->TryCancel();
-            call->Finish();
-          } else {
-            call->OnRead([](auto*, auto&& server_note) {
-              if (server_note) {
+  // RouteChat() function implements bidirectional streaming RPC. The client
+  // sends the sequence of 4 messages. Each message consists of the RouteNote
+  // data structure (you can check this data structure in the
+  // protos/route_guide.proto). The server gets this stream  of points and
+  // checks if the vector received_notes_ contains already sended points. If
+  // so, then the server prepare the stream of this points.
```

Comment:
Clarifying suggestions:
```suggestion
  // RouteChat() demonstrates a bidirectional streaming RPC. The client
  // sends a sequence of 4 messages. Each message consists of a RouteNote
  // (see protos/route_guide.proto), which contains a location. The server 
  // streams back any notes it has previously received for those locations.
```